Clarify Ray runtime usage

alex000kim opened the pull request:
Improves documentation and examples to clarify Ray runtime usage
  patterns across:
  - FAQ section with common Ray runtime questions
  - Training guide with Ray-specific guidance
  - Distributed jobs documentation
  - Ray training example with enhanced README and YAML
  configuration

andylizf commented:
Also, I was wondering how can we down the ray cluster that are launched by users **only**, without affecting the skypilot cluster management.

Michaelvll commented on docs/source/reference/faq.rst:
```diff
@@ -213,6 +213,23 @@ You will need to install a PyTorch version that is compatible with your NVIDIA d
+
+  run: |
+    if [ "$SKYPILOT_NODE_RANK" == "0" ]; then
+      ray start --head --port 6379
```
I don't think we need to add `--port` here.
```suggestion
      ray start --head
```

Michaelvll commented on docs/source/reference/faq.rst:
```diff
@@ -213,6 +213,23 @@ You will need to install a PyTorch version that is compatible with your NVIDIA d
+      ray start --head --port 6379
+      python your_script.py
+    else
+      ray start --address $HEAD_IP:6379
```
HEAD_IP is not available by SkyPilot. We need to do the normal parsing of NODE_IPS as our example yamls.

alex000kim commented:
thanks for the comments @Michaelvll . 
As I think I addressed all comments. 
Wrt:
> Could we also add how skypilot uses ray in the following page for advanced users to refer to (may need to rename the page; ok to do it in another PR)? https://docs.skypilot.co/en/latest/reference/architecture/state.html

I changed `state.rst` to `internals.rst` and added a section on "Internal Ray usage". 
Please take a look

Michaelvll commented on docs/source/reference/architecture/internals.rst:
```diff
@@ -0,0 +1,85 @@
+SkyPilot Internal Architecture
+===============================
+
+All SkyPilot internal data is stored in the ``~/.sky/`` directory of the API server (either your local machine or a remote server):
```
Make the original `SkyPilot State Management` as a subtitle here as well?

Michaelvll commented on docs/source/reference/training-guide.rst:
```diff
@@ -97,6 +97,24 @@ Use high-performance networking
+When running Ray workloads on SkyPilot:
+
+- **Running** ``ray start`` **is fine** - Ray defaults to port 6379, while SkyPilot uses port 6380 internally
+- **Never use** ``ray.init(address="auto")`` - It connects to SkyPilot's internal Ray cluster
```
If a user has run `ray start`, `ray.init()` will connect to the correct ray cluster right?

alex000kim replied:
That's right. I just ran some tests and was able to confirm this. I changed the wording and added some recommendations

alex000kim replied:
For reference, here's the test that I ran:
```
name: test-ray-simple

resources:
  cpus: 2+

setup: |
  pip install ray

run: |
  cat > test_ray.py << 'EOF'
  import ray
  import os
  
  # check if skypilot's ray is running on 6380
  os.system('ps aux | grep "raylet" | grep "6380" | grep -v grep > /dev/null && echo "skypilot ray found on 6380" || echo "no skypilot ray?"')
  
  # start our ray on 6379
  print('\nstarting user ray on 6379...')
  os.system('ray start --head --port 6379 --disable-usage-stats >/dev/null 2>&1')
  
  tests = [
      ('ray.init()', lambda: ray.init()),
      ('ray.init(address="auto")', lambda: ray.init(address='auto')),
      ('ray.init(address="localhost:6379")', lambda: ray.init(address='localhost:6379'))
  ]
  
  for name, init_func in tests:
      print(f'\n{name}:')
      try:
          ctx = init_func()
          # this is the interesting bit - which port did we connect to?
          if hasattr(ctx, 'address_info'):
              addr = ctx.address_info.get('address', 'unknown')
              print(f'  connected to {addr}')
          else:
              print('  connected (no address info)')
          ray.shutdown()
      except Exception as e:
          print(f'  failed: {e}')
  
  print('ray.init() and auto both pick user cluster when available')
  EOF
  
  python test_ray.py
  
  # cleanup our ray, leave skypilot's alone
  pkill -f "ray start --head --port 6379" || true
  ```
  `sky launch -c test-ray test_ray.yaml --infra gcp -y `

Michaelvll commented on docs/source/reference/training-guide.rst:
```diff
@@ -97,6 +97,24 @@ Use high-performance networking
+
+- **Running** ``ray start`` **is fine** - Ray defaults to port 6379, while SkyPilot uses port 6380 internally
+- **Never use** ``ray.init(address="auto")`` - It connects to SkyPilot's internal Ray cluster
+- **Start Ray head on rank 0, workers on other ranks** - See :ref:`distributed Ray example <dist-jobs>`
```
We should refer to the `distributed Ray example` at the beginning of this section. And just move these important notes to that page, as these notes are a bit hard to understand if not coming with the concrete example.

Michaelvll reviewed (commented):
Thanks @alex000kim for the update! Proposed some reorganization. : )
